src/gui/vector/qgsattributesformmodel.cpp

```diff
@@ -0,0 +1,1162 @@
+  {
+    if ( index.isValid() )
+    {
+      const QString nodeId = index.data( QgsAttributesFormModel::NodeIdRole ).toString();
+      const QString nodeName = index.data( QgsAttributesFormModel::NodeNameRole ).toString();
+      int nodeType = index.data( QgsAttributesFormModel::NodeTypeRole ).toInt();
+      const auto nodeData = index.data( QgsAttributesFormModel::NodeDataRole );
+
+      stream << nodeId << nodeType << nodeName << nodeData;
+    }
+  }
+
+  data->setData( format, encoded );
+  return data;
+}
+
+QModelIndex QgsAttributesFormModel::index( int row, int column, const QModelIndex &parent ) const
+{
+  if ( !hasIndex( row, column, parent ) )
+    return QModelIndex();
+
+  QgsAttributesFormTreeNode *parentNode = nodeForIndex( parent );
+  if ( !parentNode )
+    return QModelIndex();
+
+  if ( QgsAttributesFormTreeNode *childNode = parentNode->child( row ) )
+    return createIndex( row, column, childNode );
+
+  return QModelIndex();
+}
+
+QModelIndex QgsAttributesFormModel::parent( const QModelIndex &index ) const
+{
+  if ( !index.isValid() )
+    return QModelIndex();
+
+  QgsAttributesFormTreeNode *childNode = nodeForIndex( index );
+  QgsAttributesFormTreeNode *parentNode = childNode ? childNode->parent() : nullptr;
+
+  return ( parentNode != mRootNode.get() && parentNode != nullptr )
+           ? createIndex( parentNode->row(), 0, parentNode )
+           : QModelIndex();
+}
+
+QModelIndex QgsAttributesFormModel::firstTopMatchingModelIndex( const QgsAttributesFormTreeData::AttributesFormTreeNodeType &nodeType, const QString &nodeId ) const
+{
+  QgsAttributesFormTreeNode *node = mRootNode->firstTopChild( nodeType, nodeId );
+  return node ? createIndex( node->row(), 0, node ) : QModelIndex();
+}
+
+QModelIndex QgsAttributesFormModel::firstRecursiveMatchingModelIndex( const QgsAttributesFormTreeData::AttributesFormTreeNodeType &nodeType, const QString &nodeId ) const
+{
+  QgsAttributesFormTreeNode *node = mRootNode->firstChildRecursive( nodeType, nodeId );
+  return node ? createIndex( node->row(), 0, node ) : QModelIndex();
+}
+
+
+//
+// QgsAttributesAvailableWidgetsModel
+//
+
+QgsAttributesAvailableWidgetsModel::QgsAttributesAvailableWidgetsModel( QgsVectorLayer *layer, QgsProject *project, QObject *parent )
+  : QgsAttributesFormModel( layer, parent )
+  , mProject( project )
+{
+}
+
+Qt::ItemFlags QgsAttributesAvailableWidgetsModel::flags( const QModelIndex &index ) const
+{
+  if ( !index.isValid() )
+    return Qt::NoItemFlags;
+
+  return Qt::ItemIsEnabled | Qt::ItemIsSelectable | Qt::ItemIsDragEnabled;
+}
+
+QVariant QgsAttributesAvailableWidgetsModel::headerData( int section, Qt::Orientation orientation, int role ) const
+{
+  Q_UNUSED( section )
+  return orientation == Qt::Horizontal && role == Qt::DisplayRole ? tr( "Available Widgets" ) : QVariant {};
+}
+
+void QgsAttributesAvailableWidgetsModel::populate()
+{
+  beginResetModel();
+  mRootNode->deleteChildren();
+
+  // load Fields
```

Comment:
Please harmonize capitalization throughout this function. My preference: Load fields